vendor/bat-native-ads/src/bat/ads/internal/ml/model/linear/linear.cc

```diff
@@ -0,0 +1,69 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <algorithm>
+#include <cmath>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "bat/ads/internal/ml/data/vector_data.h"
+#include "bat/ads/internal/ml/ml_prediction_util.h"
+#include "bat/ads/internal/ml/model/linear/linear.h"
+
+namespace ads {
+namespace ml {
+namespace model {
+
+Linear::Linear() {}
+
+Linear::Linear(const std::map<std::string, VectorData>& weights,
+               const std::map<std::string, double>& biases) {
+  weights_ = weights;
+  biases_ = biases;
+}
+
+Linear::Linear(const Linear& linear_model) = default;
+
+Linear::~Linear() = default;
+
+PredictionMap Linear::Predict(const VectorData& x) const {
+  PredictionMap predictions;
+  for (const auto& kv : weights_) {
+    double prediction = kv.second * x;
+    const auto iter = biases_.find(kv.first);
+    if (iter != biases_.end()) {
+      prediction += iter->second;
+    }
+    predictions[kv.first] = prediction;
+  }
+  return predictions;
+}
+
+PredictionMap Linear::GetTopPredictions(const VectorData& x,
+                                        const int top_count) const {
+  PredictionMap prediction_map = Predict(x);
+  PredictionMap prediction_map_softmax = Softmax(prediction_map);
+  std::vector<std::pair<double, std::string>> prediction_order;
+  prediction_order.reserve(prediction_map_softmax.size());
+  for (auto iter = prediction_map_softmax.begin();

```

Comment:
Can we please use a range based for-loop. i.e. `for (const auto& item : prediction_map_softmax) {`. We should use range based for-loops where possible in other parts of this PR too

Reply:
Done.